File: 2020/Python/Day4.py
import os
import sys


def main():
    inf = os.path.dirname(os.getcwd())
    inf = os.path.join(inf, "input")
    if not os.path.exists(inf) or os.path.isdir(inf):
        print(inf + " does not exist!")
        inf += ".txt"
        print(f"trying {inf} instead.")

    if not os.path.exists(inf) or os.path.isdir(inf):
        print(inf + " does not exist!")
        inf = os.path.join(os.path.dirname(inf), "input")
        inf = os.path.join(inf, "Day4.txt")
        print(f"trying {inf} instead.")

    if not os.path.exists(inf) or os.path.isdir(inf):
        print("None of the expected inputs exist!", file=sys.stderr)
        return
    else:
        print(f"Using input file {inf}.")

    inputFile = open(inf, 'r')
    requiredFields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"];
    validEyeColors = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
    validPassports1 = 0
    validPassports2 = 0
    fields = {}
    for line in inputFile.readlines():
        line = line.rstrip()
        if len(line) <= 1:
            invalid1 = False
            invalid2 = False
            for field in requiredFields:
                if field in fields:
                    if field == "byr":
                        byr = int(fields[field])
                        if byr < 1920 or byr > 2002:
                            invalid2 = True
                    elif field == "iyr":
                        iyr = int(fields[field])
                        if iyr < 2010 or iyr > 2020:
                            invalid2 = True
                    elif field == "eyr":
                        eyr = int(fields[field])
                        if eyr < 2020 or eyr > 2030:
                            invalid2 = True
                    elif field == "hgt":
                        hgt = fields[field]
                        if len(hgt) < 4:
                            invalid2 = True
                            continue
                        unit = hgt[-2:]
                        value = int(hgt[:-2])
                        if unit == "cm":
                            if value < 150 or value > 193:
                                invalid2 = True
                        elif unit == "in":
                            if value < 59 or value > 76:
                                invalid2 = True
                    elif field == "hcl":
                        hcl = fields[field]
                        if hcl[0] != '#' or len(hcl) != 7:
                            invalid2 = True
                            continue
                        try:
                            int(hcl[1:], 16)
                        except ValueError:
                            invalid2 = True
                    elif field == "ecl":
                        if not fields[field] in validEyeColors:
                            invalid2 = True
                    elif field == "pid":
                        pid = fields[field]
                        if len(pid) != 9:
                            invalid2 = True
                            continue
                        try:
                            int(pid)
                        except ValueError:
                            invalid2 = True
                            
                else:
                    invalid1 = invalid2 = True
                    break;

            if not invalid1:
                validPassports1 += 1
            if not invalid2:
                validPassports2 += 1
            fields.clear()
        else:
            for field in line.split(" "):
                fields[field.split(":")[0]] = field.split(":")[1]

    print(f"Step 1 found {validPassports1} valid passports.")
    print(f"Step 2 found {validPassports2} valid passports.")


if __name__ == '__main__':
    main()

File: 2020/Java/src/aoc/Day4.java
package aoc;

import java.io.IOException;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public abstract class Day4 {

	public static void main(String[] args) throws IOException {

		final String[] requiredFields = new String[] { "byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid" };
		final String[] validEyeColors = new String[] { "amb", "blu", "brn", "gry", "grn", "hzl", "oth" };
		int validPassports1 = 0;
		int validPassports2 = 0;
		Map<String, String> fields = new HashMap<>();
		for (String line : InputReader.readInputFileLines(4)) {
			if (line.isEmpty()) {
				boolean invalid1 = false;
				boolean invalid2 = false;
				for (String field : requiredFields) {
					if (fields.containsKey(field)) {
						switch (field) {
						case "byr":
							int byr = Integer.parseInt(fields.get(field));
							if (byr < 1920 || byr > 2002)
								invalid2 = true;
							break;
						case "iyr":
							int iyr = Integer.parseInt(fields.get(field));
							if (iyr < 2010 || iyr > 2020)
								invalid2 = true;
							break;
						case "eyr":
							int eyr = Integer.parseInt(fields.get(field));
							if (eyr < 2020 || eyr > 2030)
								invalid2 = true;
							break;
						case "hgt":
							String hgt = fields.get(field);
							if (hgt.length() < 4) {
								invalid2 = true;
								break;
							}
							String unit = hgt.substring(hgt.length() - 2, hgt.length());
							int value = Integer.parseInt(hgt.substring(0, hgt.length() - 2));
							if (unit.equals("cm")) {
								if (value < 150 || value > 193)
									invalid2 = true;
							} else if (unit.equals("in")) {
								if (value < 59 || value > 76)
									invalid2 = true;
							}
							break;
						case "hcl":
							String hcl = fields.get(field);
							if (hcl.charAt(0) != '#' || hcl.length() != 7) {
								invalid2 = true;
								break;
							}
							try {
								Integer.parseInt(hcl.substring(1), 16);
							} catch (NumberFormatException e) {
								invalid2 = true;
							}
							break;
						case "ecl":
							String ecl = fields.get(field);
							if (!Arrays.stream(validEyeColors).anyMatch(ecl::equals))
								invalid2 = true;
							break;
						case "pid":
							String pid = fields.get(field);
							if (pid.length() != 9) {
								invalid2 = true;
								break;
							}
							try {
								Integer.parseInt(pid);
							} catch (NumberFormatException e) {
								invalid2 = true;
							}
							break;
						}
					} else {
						invalid1 = invalid2 = true;
					}
				}

				if (!invalid1) {
					validPassports1++;
				}
				if (!invalid2) {
					validPassports2++;
				}
				fields.clear();
			} else {
				for (String field : line.split(" ")) {
					fields.put(field.substring(0, field.indexOf(':')), field.substring(field.indexOf(':') + 1));
				}
			}
		}

		System.out.printf("Step 1 found %d valid passports.%n", validPassports1);
		System.out.printf("Step 2 found %d valid passports.%n", validPassports2);
	}

}
